File: flutter/ai_flutter_agent/lib/src/result_handle_store.dart
import 'dart:math';

class ResultHandleStore {
  ResultHandleStore({this.maxHandles = 96});

  final int maxHandles;
  final Map<String, _ResultHandleEntry> _entries = <String, _ResultHandleEntry>{};
  final List<String> _order = <String>[];

  Map<String, dynamic> buildItemsResult({
    required String sourceToolId,
    required String fieldName,
    required List<Object?> items,
    required Map<String, dynamic> summary,
    int pageSize = 20,
  }) {
    final safePageSize = pageSize <= 0 ? 20 : pageSize.clamp(1, 200).toInt();
    final returned = min(safePageSize, items.length);
    final hasMore = returned < items.length;
    final response = <String, dynamic>{
      'summary': summary,
      'returned': returned,
      'pageSize': safePageSize,
      'total': items.length,
      'hasMore': hasMore,
      fieldName: items.take(returned).toList(),
    };

    if (hasMore) {
      final handleId = _createItemsHandle(
        sourceToolId: sourceToolId,
        fieldName: fieldName,
        items: items,
        summary: summary,
      );
      response['resultHandle'] = handleId;
    }

    return response;
  }

  Map<String, dynamic> buildTextResult({
    required String sourceToolId,
    required String text,
    required Map<String, dynamic> summary,
    int offset = 0,
    int length = 4096,
  }) {
    final safeOffset = offset.clamp(0, text.length).toInt();
    final safeLength = length <= 0 ? 4096 : length.clamp(1, 32768).toInt();
    final count = min(safeLength, text.length - safeOffset);
    final hasMore = safeOffset + count < text.length;
    final chunk = count <= 0 ? '' : text.substring(safeOffset, safeOffset + count);

    final response = <String, dynamic>{
      'summary': summary,
      'offset': safeOffset,
      'limit': safeLength,
      'count': count,
      'totalChars': text.length,
      'hasMore': hasMore,
      'content': chunk,
    };

    if (hasMore) {
      final handleId = _createTextHandle(
        sourceToolId: sourceToolId,
        summary: summary,
        text: text,
      );
      response['resultHandle'] = handleId;
    }

    return response;
  }

  Map<String, dynamic>? buildPage(String handleId, {int offset = 0, int limit = 0}) {
    final entry = _entries[handleId];
    if (entry == null) {
      return null;
    }

    if (entry.kind == _ResultHandleKind.text) {
      return _buildTextPage(entry, offset: offset, limit: limit);
    }
    return _buildItemsPage(entry, offset: offset, limit: limit);
  }

  String _createItemsHandle({
    required String sourceToolId,
    required String fieldName,
    required List<Object?> items,
    required Map<String, dynamic> summary,
  }) {
    final entry = _ResultHandleEntry(
      id: _newHandleId(),
      kind: _ResultHandleKind.items,
      sourceToolId: sourceToolId,
      fieldName: fieldName,
      createdAt: DateTime.now().toUtc().toIso8601String(),
      summary: summary,
      items: List<Object?>.from(items),
    );
    _add(entry);
    return entry.id;
  }

  String _createTextHandle({
    required String sourceToolId,
    required Map<String, dynamic> summary,
    required String text,
  }) {
    final entry = _ResultHandleEntry(
      id: _newHandleId(),
      kind: _ResultHandleKind.text,
      sourceToolId: sourceToolId,
      fieldName: 'content',
      createdAt: DateTime.now().toUtc().toIso8601String(),
      summary: summary,
      text: text,
    );
    _add(entry);
    return entry.id;
  }

  void _add(_ResultHandleEntry entry) {
    _entries[entry.id] = entry;
    _order.add(entry.id);
    while (_order.length > maxHandles) {
      final oldest = _order.removeAt(0);
      _entries.remove(oldest);
    }
  }

  Map<String, dynamic> _buildItemsPage(
    _ResultHandleEntry entry, {
    required int offset,
    required int limit,
  }) {
    final items = entry.items;
    final safeOffset = offset.clamp(0, items.length).toInt();
    final safeLimit = limit <= 0 ? 20 : limit.clamp(1, 200).toInt();
    final count = min(safeLimit, items.length - safeOffset);
    final hasMore = safeOffset + count < items.length;
    return <String, dynamic>{
      'ok': true,
      'handleId': entry.id,
      'kind': 'items',
      'sourceToolId': entry.sourceToolId,
      'fieldName': entry.fieldName,
      'createdAt': entry.createdAt,
      'offset': safeOffset,
      'limit': safeLimit,
      'count': count,
      'total': items.length,
      'hasMore': hasMore,
      'summary': entry.summary,
      entry.fieldName: items.skip(safeOffset).take(count).toList(),
    };
  }

  Map<String, dynamic> _buildTextPage(
    _ResultHandleEntry entry, {
    required int offset,
    required int limit,
  }) {
    final text = entry.text;
    final safeOffset = offset.clamp(0, text.length).toInt();
    final safeLimit = limit <= 0 ? 4096 : limit.clamp(1, 32768).toInt();
    final count = min(safeLimit, text.length - safeOffset);
    final hasMore = safeOffset + count < text.length;
    return <String, dynamic>{
      'ok': true,
      'handleId': entry.id,
      'kind': 'text',
      'sourceToolId': entry.sourceToolId,
      'createdAt': entry.createdAt,
      'offset': safeOffset,
      'limit': safeLimit,
      'count': count,
      'totalChars': text.length,
      'hasMore': hasMore,
      'summary': entry.summary,
      'content': count <= 0 ? '' : text.substring(safeOffset, safeOffset + count),
    };
  }

  String _newHandleId() {
    final random = Random.secure();
    final timestamp = DateTime.now().microsecondsSinceEpoch.toRadixString(16);
    final suffix = random.nextInt(0x7fffffff).toRadixString(16).padLeft(8, '0');
    return '$timestamp$suffix';
  }
}

enum _ResultHandleKind { items, text }

class _ResultHandleEntry {
  _ResultHandleEntry({
    required this.id,
    required this.kind,
    required this.sourceToolId,
    required this.fieldName,
    required this.createdAt,
    required this.summary,
    this.items = const <Object?>[],
    this.text = '',
  });

  final String id;
  final _ResultHandleKind kind;
  final String sourceToolId;
  final String fieldName;
  final String createdAt;
  final Map<String, dynamic> summary;
  final List<Object?> items;
  final String text;
}

File: vscode/ai_vscode_agent/lib/result_handle_store.js
const crypto = require('node:crypto');

const { clampNumber } = require('./models');

class ResultHandleStore {
  constructor(maxHandles = 96) {
    this.maxHandles = maxHandles;
    this.entries = new Map();
    this.order = [];
  }

  buildItemsResult({ sourceToolId, fieldName, items, summary, pageSize = 20 }) {
    const safePageSize = clampNumber(pageSize || 20, 1, 200);
    const returned = Math.min(safePageSize, items.length);
    const hasMore = returned < items.length;
    const response = {
      summary,
      returned,
      pageSize: safePageSize,
      total: items.length,
      hasMore,
      [fieldName]: items.slice(0, returned),
    };
    if (hasMore) {
      response.resultHandle = this.#createItemsHandle({ sourceToolId, fieldName, items, summary });
    }
    return response;
  }

  buildTextResult({ sourceToolId, text, summary, offset = 0, length = 4096 }) {
    const safeOffset = clampNumber(offset || 0, 0, text.length);
    const safeLength = clampNumber(length || 4096, 1, 32768);
    const count = Math.min(safeLength, text.length - safeOffset);
    const hasMore = safeOffset + count < text.length;
    const response = {
      summary,
      offset: safeOffset,
      limit: safeLength,
      count,
      totalChars: text.length,
      hasMore,
      content: count > 0 ? text.slice(safeOffset, safeOffset + count) : '',
    };
    if (hasMore) {
      response.resultHandle = this.#createTextHandle({ sourceToolId, summary, text });
    }
    return response;
  }

  buildPage(handleId, { offset = 0, limit = 0 } = {}) {
    const entry = this.entries.get(handleId);
    if (!entry) {
      return null;
    }
    return entry.kind === 'text'
      ? this.#buildTextPage(entry, { offset, limit })
      : this.#buildItemsPage(entry, { offset, limit });
  }

  #createItemsHandle({ sourceToolId, fieldName, items, summary }) {
    const entry = {
      id: this.#newHandleId(),
      kind: 'items',
      sourceToolId,
      fieldName,
      createdAt: new Date().toISOString(),
      summary,
      items,
    };
    this.#add(entry);
    return entry.id;
  }

  #createTextHandle({ sourceToolId, summary, text }) {
    const entry = {
      id: this.#newHandleId(),
      kind: 'text',
      sourceToolId,
      fieldName: 'content',
      createdAt: new Date().toISOString(),
      summary,
      text,
    };
    this.#add(entry);
    return entry.id;
  }

  #add(entry) {
    this.entries.set(entry.id, entry);
    this.order.push(entry.id);
    while (this.order.length > this.maxHandles) {
      const oldest = this.order.shift();
      this.entries.delete(oldest);
    }
  }

  #buildItemsPage(entry, { offset, limit }) {
    const safeOffset = clampNumber(offset || 0, 0, entry.items.length);
    const safeLimit = clampNumber(limit || 20, 1, 200);
    const count = Math.min(safeLimit, entry.items.length - safeOffset);
    return {
      ok: true,
      handleId: entry.id,
      kind: 'items',
      sourceToolId: entry.sourceToolId,
      fieldName: entry.fieldName,
      createdAt: entry.createdAt,
      offset: safeOffset,
      limit: safeLimit,
      count,
      total: entry.items.length,
      hasMore: safeOffset + count < entry.items.length,
      summary: entry.summary,
      [entry.fieldName]: entry.items.slice(safeOffset, safeOffset + count),
    };
  }

  #buildTextPage(entry, { offset, limit }) {
    const safeOffset = clampNumber(offset || 0, 0, entry.text.length);
    const safeLimit = clampNumber(limit || 4096, 1, 32768);
    const count = Math.min(safeLimit, entry.text.length - safeOffset);
    return {
      ok: true,
      handleId: entry.id,
      kind: 'text',
      sourceToolId: entry.sourceToolId,
      createdAt: entry.createdAt,
      offset: safeOffset,
      limit: safeLimit,
      count,
      totalChars: entry.text.length,
      hasMore: safeOffset + count < entry.text.length,
      summary: entry.summary,
      content: count > 0 ? entry.text.slice(safeOffset, safeOffset + count) : '',
    };
  }

  #newHandleId() {
    return `${Date.now().toString(16)}${crypto.randomBytes(4).toString('hex')}`;
  }
}

module.exports = {
  ResultHandleStore,
};
